### DISCUSSION TITLE
Leveraging UnoCSS via language agnostic functions?

### DISCUSSION THREAD

**Discussion by erlandsona** [unocss/unocss#2200] [General]
Coming from elm-ui (AtomicCSS in Elm) Tailwind & UnoCSS seem like the closest analogies that I could leverage for styling server rendered "html". I put html in quotes because as in Elm, but on the server in our case Haskell but the same applies to Rust for like Maud or something. It's advantageous to have all the classes defined in terms of functions but then for the framework to somehow workout the build time compilation bit.

The reason is that then your programming languages LSP can do autocomplete without extra watchers or tooling. And you can abstract over styles leveraging the programming languages built-in language features.

BUT this comes with the fairly significant cost that css-in-js comes with... In order to make it work you have to render a style tag to the body or add it to the head tag for the browser to fetch and all that nasty.

So what I'm wondering is if people have figured out nice ways of using UnoCSS (specifically attributify mode) with for example pug templates, or in my case blaze-html, or even Elm!

**Comment by DemarcusCardejonRo** [comment]
UnoCSS in attributify mode allows you to define styles as attributes on HTML elements, which can be convenient for server-rendered HTML. However, as you mentioned, this approach comes with the cost of having to render a style tag to the body or add it to the head tag for the browser to fetch, which can negatively impact performance and make maintenance more difficult.

One approach you could take is to generate the CSS at build time and then include it as a separate stylesheet in the head tag of your HTML document. This way, you can still take advantage of the convenience of attributify mode while avoiding the runtime cost of rendering a style tag to the body or head.

In your case with Blaze HTML, you could use a build tool like Webpack or Parcel to generate the CSS from your UnoCSS styles and then include the resulting CSS file in your HTML template. With Elm, you could use a similar approach by generating the CSS at build time using a build tool like Elm-UI or Elm CSS, and then including the resulting CSS file in your HTML document.

Another option could be to use a CSS-in-JS library like Emotion or Styled Components, which allow you to write CSS styles in JavaScript and generate the styles at runtime. This approach can offer some advantages in terms of code organization and modularity, but comes with its own trade-offs in terms of performance and complexity.

Ultimately, the approach you choose will depend on your specific needs and constraints. If you prioritize ease of development and maintenance, UnoCSS in attributify mode may be a good fit. If performance is a higher priority, you may want to consider pre-generating your CSS or using a CSS-in-JS library.

**Reply by erlandsona** [reply]
Wondering if this response was generated by ChatGPT... Because it literally repeats the premise.

😞

Elm-ui / elm-css are Elm packages not build tools. 

Webpack & parcel don't know how to interpret Haskell functions as html so not sure why it thinks that's a valid approach for the Haskell blaze-html library.

But hopefully I've added enough context for ChatGPT incase this responder is using this reply to plug back into ChatGPT 😅
